File: frontend/src/constants/promptExpander.ts
// Prompt Expander の定数（バックエンド consts/prompt_expander.py のミラー）

import { isV5ImageModel } from "./novelaiImageModels";

/** Prompt Expander で選択できる NovelAI 画像モデル（表示順） */
export const PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS = [
  "nai-diffusion-5-full",
  "nai-diffusion-5-curated",
  "nai-diffusion-4-5-full",
  "nai-diffusion-4-5-curated",
] as const;

export type PromptExpanderImageModel =
  (typeof PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS)[number];

export const DEFAULT_PROMPT_EXPANDER_IMAGE_MODEL: PromptExpanderImageModel =
  "nai-diffusion-4-5-full";

/** モデル選択肢の表示ラベル */
export const PROMPT_EXPANDER_IMAGE_MODEL_LABELS: Record<
  PromptExpanderImageModel,
  string
> = {
  "nai-diffusion-5-full": "NAI Diffusion V5 Full",
  "nai-diffusion-5-curated": "NAI Diffusion V5 Curated",
  "nai-diffusion-4-5-full": "NAI Diffusion V4.5 Full",
  "nai-diffusion-4-5-curated": "NAI Diffusion V4.5 Curated",
};

/** バッジ等で使う短いモデル表記 */
export const PROMPT_EXPANDER_IMAGE_MODEL_SHORT_LABELS: Record<
  PromptExpanderImageModel,
  string
> = {
  "nai-diffusion-5-full": "V5 Full",
  "nai-diffusion-5-curated": "V5 Curated",
  "nai-diffusion-4-5-full": "V4.5 Full",
  "nai-diffusion-4-5-curated": "V4.5 Curated",
};

/** キャラクタープロンプトの上限数（V5 系 / V4.5 系） */
export const MAX_CHARACTER_PROMPTS_V5 = 22;
export const MAX_CHARACTER_PROMPTS_V45 = 6;

/** 画像モデルごとのキャラクタープロンプト上限数を返す */
export function getMaxCharacterPrompts(
  model: string | null | undefined,
): number {
  return isV5ImageModel(model)
    ? MAX_CHARACTER_PROMPTS_V5
    : MAX_CHARACTER_PROMPTS_V45;
}

/** 画像サイズの選択肢 */
export const PROMPT_EXPANDER_IMAGE_SIZES = [
  "portrait",
  "landscape",
  "square",
] as const;

export type PromptExpanderImageSize =
  (typeof PROMPT_EXPANDER_IMAGE_SIZES)[number];

/** プロンプト拡張に使える NovelAI テキストモデル */
export const NOVELAI_TEXT_MODEL_OPTIONS = ["glm-4-6", "xialong-v1"] as const;

export type NovelaiTextModel = (typeof NOVELAI_TEXT_MODEL_OPTIONS)[number];

/** 拡張モード（日本語プロンプト / Danbooru タグ） */
export type PromptExpandMode = "japanese" | "tags";

/** 生成元画像の種別 */
export type PromptExpanderSourceKind = "none" | "history" | "entry" | "upload";

/** i2i 強度の入力範囲 */
export const PROMPT_EXPANDER_I2I_STRENGTH_MIN = 0.05;
export const PROMPT_EXPANDER_I2I_STRENGTH_MAX = 0.99;
/** i2i ノイズの入力範囲 */
export const PROMPT_EXPANDER_I2I_NOISE_MIN = 0;
export const PROMPT_EXPANDER_I2I_NOISE_MAX = 0.5;
/** シードの入力範囲 */
export const PROMPT_EXPANDER_SEED_MAX = 999999999;
/** PE ローカルメモリの最大文字数 */
export const PROMPT_EXPANDER_MEMORY_MAX_LENGTH = 10000;

/** 漫画モード: コマ数（0 = おまかせ） */
export const PROMPT_EXPANDER_MANGA_PANEL_COUNT_AUTO = 0;
export const PROMPT_EXPANDER_MANGA_PANEL_COUNT_MIN = 1;
export const PROMPT_EXPANDER_MANGA_PANEL_COUNT_MAX = 6;
/** 漫画モード: コマ割り */
export const PROMPT_EXPANDER_MANGA_LAYOUTS = [
  "auto",
  "vertical",
  "horizontal",
  "grid",
] as const;
export type PromptExpanderMangaLayout =
  (typeof PROMPT_EXPANDER_MANGA_LAYOUTS)[number];
/** 漫画モード: セリフ・効果音の言語（auto = 指示文の言語に合わせる） */
export const PROMPT_EXPANDER_MANGA_TEXT_LANGUAGES = [
  "auto",
  "ja",
  "en",
] as const;
export type PromptExpanderMangaTextLanguage =
  (typeof PROMPT_EXPANDER_MANGA_TEXT_LANGUAGES)[number];
/** 漫画モード: 読み順（rtl = 日本式: 右上始まりで右→左・上→下、ltr = 西洋式） */
export const PROMPT_EXPANDER_MANGA_READING_DIRECTIONS = ["rtl", "ltr"] as const;
export type PromptExpanderMangaReadingDirection =
  (typeof PROMPT_EXPANDER_MANGA_READING_DIRECTIONS)[number];

/** 漫画モードを使える画像モデルか（V5 系のみ） */
export function supportsMangaMode(model: string | null | undefined): boolean {
  return isV5ImageModel(model);
}

/** 精密参照（NovelAI character reference）の種別 */
export const PROMPT_EXPANDER_REFERENCE_TYPES = [
  "character",
  "style",
  "character&style",
] as const;
export type PromptExpanderReferenceType =
  (typeof PROMPT_EXPANDER_REFERENCE_TYPES)[number];
/** 既定は Adventure の立ち絵生成と同じ（立ち絵差分では同一性の固定が目的） */
export const DEFAULT_PROMPT_EXPANDER_REFERENCE_TYPE: PromptExpanderReferenceType =
  "character";
export const DEFAULT_PROMPT_EXPANDER_REFERENCE_STRENGTH = 0.85;
export const DEFAULT_PROMPT_EXPANDER_REFERENCE_FIDELITY = 1;
/** 精密参照 1 枚あたりの Anlas 消費（設定応答の anlas_per_reference が無いときの既定） */
export const PROMPT_EXPANDER_ANLAS_PER_REFERENCE = 5;
/** 精密参照の Anlas 確認ダイアログを抑止する sessionStorage キー（ブラウザセッション単位） */
export const PROMPT_EXPANDER_ANLAS_WARN_SUPPRESSED_KEY =
  "prompt_expander_anlas_warn_suppressed";

/** 精密参照を使える画像モデルか（V4.5 系のみ。V5 は API 非対応） */
export function supportsPreciseReference(
  model: string | null | undefined,
): boolean {
  return (
    (PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS as readonly string[]).includes(
      model ?? "",
    ) && !isV5ImageModel(model)
  );
}

/**
 * 背景透過をプロンプト指示だけでネイティブに行えるモデルか（V5 系）。
 * V4.5 系は白背景で生成し、表示時にフロントで切り抜く。
 */
export function usesNativeTransparency(
  model: string | null | undefined,
): boolean {
  return isV5ImageModel(model);
}

/** 参照種別の i18n キー（"character&style" は "&" をキーに使えないため characterStyle に写す） */
export function referenceTypeI18nKey(
  type: PromptExpanderReferenceType | string,
): string {
  return type === "character&style" ? "characterStyle" : type;
}

/**
 * 背景透過タグの強調段数（V4.5 系のみ効く）。
 * NovelAI の {} 記法は 1 段ごとに重み 1.05 倍で、無強調だと白背景の指定が
 * 無視されて背景が描かれ、切り抜きが失敗することがあるため既定は 2 段。
 * V5 はネイティブ透過なので段数を指定しても効かない。
 */
export const PROMPT_EXPANDER_TRANSPARENT_EMPHASIS_LEVELS = [
  0, 1, 2, 3,
] as const;
export type PromptExpanderTransparentEmphasis =
  (typeof PROMPT_EXPANDER_TRANSPARENT_EMPHASIS_LEVELS)[number];
export const DEFAULT_PROMPT_EXPANDER_TRANSPARENT_EMPHASIS: PromptExpanderTransparentEmphasis = 2;

/**
 * 背景透過で送信プロンプトの末尾へ足されるタグ（backend consts のミラー）。
 * エントリの最終プロンプトには保存されないため、欄の下のプレビューで見せる。
 */
export const TRANSPARENT_BACKGROUND_TAGS_V5 = [
  "transparent background",
  "no shadow",
] as const;
export const TRANSPARENT_BACKGROUND_TAGS_V45 = [
  "simple background",
  "white background",
  "no shadow",
] as const;
export const TRANSPARENT_BACKGROUND_NEGATIVE_TAGS = [
  "multiple views",
  "reference sheet",
  "character sheet",
  "turnaround",
] as const;
/** 強調対象は背景そのものを決めるタグだけ（no shadow は素のまま） */
export const TRANSPARENT_BACKGROUND_EMPHASIZED_TAGS: ReadonlySet<string> =
  new Set(["simple background", "white background"]);

/** NovelAI の {} 記法でタグを強調する（level 0 は素通し） */
export function emphasizeTag(tag: string, level: number): string {
  return level <= 0 ? tag : "{".repeat(level) + tag + "}".repeat(level);
}

/** 背景透過で正プロンプトへ足すタグ。強調は V4.5 の背景タグにだけ効く */
export function transparentBackgroundTags(
  model: string | null | undefined,
  emphasis = 0,
): string[] {
  if (usesNativeTransparency(model)) return [...TRANSPARENT_BACKGROUND_TAGS_V5];
  const level = Math.max(0, emphasis);
  return TRANSPARENT_BACKGROUND_TAGS_V45.map((tag) =>
    TRANSPARENT_BACKGROUND_EMPHASIZED_TAGS.has(tag)
      ? emphasizeTag(tag, level)
      : tag,
  );
}

/** 重み記法や括弧を外して照合用の表記へ整える（backend の normalize_tag_for_match と同じ） */
export function normalizeTagForMatch(tag: string): string {
  return tag
    .toLowerCase()
    .replace(/\d+(?:\.\d+)?::/g, "")
    .replace(/::/g, " ")
    .replace(/[{}[\]()]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * プロンプトへ実際に追加されるタグだけを返す（backend の merge_tags と同じ判定）。
 * 既に同じタグが入力済みなら追加されないので、プレビューにも出さない。
 */
export function appendedTags(
  prompt: string,
  tags: readonly string[],
): string[] {
  const existing = new Set(
    prompt
      .split(",")
      .map((item) => normalizeTagForMatch(item))
      .filter((item) => item.length > 0),
  );
  return tags.filter((tag) => !existing.has(normalizeTagForMatch(tag)));
}

/** 強調段数のプレビュー表記（例: 2 -> "{{ }}"） */
export function transparentEmphasisSample(level: number): string {
  return level <= 0 ? "—" : `${"{".repeat(level)} ${"}".repeat(level)}`;
}

/**
 * インペイント（部分修正）。i2i 元をベース画像として、マスクで塗った領域だけ描き直す。
 * NovelAI はマスクを 1/8 解像度で扱うため、書き出しのグリッドもそれに合わせる
 * （固定値にすると landscape / square でマスクの縦横比が崩れる）。
 */
export const PROMPT_EXPANDER_MASK_GRID_DIVISOR = 8;
export const PROMPT_EXPANDER_BRUSH_SIZE_MIN = 4;
export const PROMPT_EXPANDER_BRUSH_SIZE_MAX = 96;
export const DEFAULT_PROMPT_EXPANDER_BRUSH_SIZE = 32;

/** V4.5 の白背景画像を切り抜くときの許容差（Adventure の立ち絵と同じ。生成画像の白はわずかに灰色に振れる） */
export const PROMPT_EXPANDER_ALPHA_OPTIONS = {
  threshold: 12,
  featherRadius: 1.8,
} as const;

/** 漫画モードで選べるコマ数の一覧（0 = おまかせ を先頭に） */
export function mangaPanelCountOptions(): number[] {
  const counts: number[] = [PROMPT_EXPANDER_MANGA_PANEL_COUNT_AUTO];
  for (
    let n = PROMPT_EXPANDER_MANGA_PANEL_COUNT_MIN;
    n <= PROMPT_EXPANDER_MANGA_PANEL_COUNT_MAX;
    n += 1
  ) {
    counts.push(n);
  }
  return counts;
}

/** 画像モデル名から表示ラベルを返す（未知名はそのまま返す） */
export function getPromptExpanderImageModelLabel(model: string): string {
  return (
    PROMPT_EXPANDER_IMAGE_MODEL_LABELS[model as PromptExpanderImageModel] ??
    model
  );
}

/** 画像モデル名から短い表示ラベルを返す（未知名はそのまま返す） */
export function getPromptExpanderImageModelShortLabel(model: string): string {
  return (
    PROMPT_EXPANDER_IMAGE_MODEL_SHORT_LABELS[
      model as PromptExpanderImageModel
    ] ?? model
  );
}

File: backend/gateway/consts/prompt_expander.py
"""Prompt Expander の境界値・選択肢（唯一の情報源）。

フロントエンドは frontend/src/constants/promptExpander.ts に同値をミラーする。
"""

from __future__ import annotations

from typing import Final

from .novelai_models import is_v5_image_model

# 画像モデル（NovelAI のみ。NSFW は family から導出する）
PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS: Final[tuple[str, ...]] = (
    "nai-diffusion-5-full",
    "nai-diffusion-5-curated",
    "nai-diffusion-4-5-full",
    "nai-diffusion-4-5-curated",
)
DEFAULT_PROMPT_EXPANDER_IMAGE_MODEL: Final[str] = "nai-diffusion-4-5-full"

# キャラクタープロンプトの上限（V5 は公式発表の 22 人、V4.5 は従来どおり 6 人）
MAX_CHARACTER_PROMPTS_V5: Final[int] = 22
MAX_CHARACTER_PROMPTS_V45: Final[int] = 6

# 無料枠で生成できるサイズのみ（large 系は Anlas を消費するため対象外）
PROMPT_EXPANDER_IMAGE_SIZES: Final[tuple[str, ...]] = (
    "portrait",
    "landscape",
    "square",
)
DEFAULT_PROMPT_EXPANDER_IMAGE_SIZE: Final[str] = "portrait"

PROMPT_EXPANDER_EXPAND_MODES: Final[tuple[str, ...]] = ("off", "japanese", "tags")
PROMPT_EXPANDER_SOURCE_KINDS: Final[tuple[str, ...]] = (
    "none",
    "history",
    "entry",
    "upload",
)

PROMPT_EXPANDER_SUGGESTION_COUNT_DEFAULT: Final[int] = 3
PROMPT_EXPANDER_SUGGESTION_COUNT_MIN: Final[int] = 1
PROMPT_EXPANDER_SUGGESTION_COUNT_MAX: Final[int] = 5

PROMPT_EXPANDER_TITLE_MAX_LEN: Final[int] = 120
PROMPT_EXPANDER_INSTRUCTION_MAX_LEN: Final[int] = 4000
PROMPT_EXPANDER_PROMPT_MAX_LEN: Final[int] = 8000
PROMPT_EXPANDER_NEGATIVE_MAX_LEN: Final[int] = 4000
PROMPT_EXPANDER_CHARACTER_PROMPT_MAX_LEN: Final[int] = 2000
PROMPT_EXPANDER_MEMORY_MAX_LEN: Final[int] = 10000
# base64 文字列としての上限（約 12MB の画像に相当）
PROMPT_EXPANDER_UPLOAD_MAX_BASE64_LEN: Final[int] = 16 * 1024 * 1024

DEFAULT_PROMPT_EXPANDER_I2I_STRENGTH: Final[float] = 0.7
DEFAULT_PROMPT_EXPANDER_I2I_NOISE: Final[float] = 0.0

# インペイント（部分修正）。i2i 元をベース画像に使い、マスクで塗った領域だけを描き直す。
# 強度・ノイズは i2i の値をそのまま inpaintImg2ImgStrength / noise として使う。
# NovelAI はマスクを 1/8 解像度で扱うため、書き出し・正規化のグリッドもそれに合わせる。
PROMPT_EXPANDER_MASK_GRID_DIVISOR: Final[int] = 8
PROMPT_EXPANDER_BRUSH_SIZE_MIN: Final[int] = 4
PROMPT_EXPANDER_BRUSH_SIZE_MAX: Final[int] = 96
DEFAULT_PROMPT_EXPANDER_BRUSH_SIZE: Final[int] = 32

# 精密参照（NovelAI character reference）。V4.5 系のみ対応で、1 枚あたり Anlas を消費する。
# 立ち絵差分では同一性の固定が目的なので、既定強度は Adventure の立ち絵生成と同じにする
PROMPT_EXPANDER_REFERENCE_TYPES: Final[tuple[str, ...]] = (
    "character",
    "style",
    "character&style",
)
DEFAULT_PROMPT_EXPANDER_REFERENCE_TYPE: Final[str] = "character"
DEFAULT_PROMPT_EXPANDER_REFERENCE_STRENGTH: Final[float] = 0.85
DEFAULT_PROMPT_EXPANDER_REFERENCE_FIDELITY: Final[float] = 1.0
PROMPT_EXPANDER_ANLAS_PER_REFERENCE: Final[int] = 5

# 背景透過。V5 系はプロンプト指示でネイティブ透過 PNG を返し、V4.5 系は白背景で生成して
# フロントの切り抜き処理（imageAlpha.ts）で透過にする。negative は複数ビュー・キャラクター
# シート化を抑える語だけに留め、複数人を禁じる語は入れない（複数キャラの透過生成も許すため）
TRANSPARENT_BACKGROUND_TAGS_V5: Final[tuple[str, ...]] = (
    "transparent background",
    "no shadow",
)
TRANSPARENT_BACKGROUND_TAGS_V45: Final[tuple[str, ...]] = (
    "simple background",
    "white background",
    "no shadow",
)
TRANSPARENT_BACKGROUND_NEGATIVE_TAGS: Final[tuple[str, ...]] = (
    "multiple views",
    "reference sheet",
    "character sheet",
    "turnaround",
)

# 背景透過タグの強調（V4.5 系のみ）。NovelAI の {} 記法は 1 段ごとに重み 1.05 倍で、
# V4/V4.5 でも有効。無強調だと白背景の指定がモデルに無視されて背景が描かれ、
# フロントの切り抜き（imageAlpha.ts）が失敗することがあるため既定で 2 段掛ける。
# V5 はネイティブ透過 PNG を返すので強調しない。
PROMPT_EXPANDER_TRANSPARENT_EMPHASIS_LEVELS: Final[tuple[int, ...]] = (0, 1, 2, 3)
DEFAULT_PROMPT_EXPANDER_TRANSPARENT_EMPHASIS: Final[int] = 2
# 強調対象は背景そのものを決めるタグだけ。no shadow は素のまま置く
TRANSPARENT_BACKGROUND_EMPHASIZED_TAGS: Final[frozenset[str]] = frozenset(
    {"simple background", "white background"}
)

# 漫画モード（NovelAI Diffusion V5 のコマ割り・吹き出し生成を LLM 拡張で支援する）
# panel_count は 0 が「おまかせ」（LLM が指示文から 2〜4 コマを選ぶ）
PROMPT_EXPANDER_MANGA_PANEL_COUNT_AUTO: Final[int] = 0
PROMPT_EXPANDER_MANGA_PANEL_COUNT_MIN: Final[int] = 1
PROMPT_EXPANDER_MANGA_PANEL_COUNT_MAX: Final[int] = 6
PROMPT_EXPANDER_MANGA_LAYOUTS: Final[tuple[str, ...]] = (
    "auto",
    "vertical",
    "horizontal",
    "grid",
)
DEFAULT_PROMPT_EXPANDER_MANGA_LAYOUT: Final[str] = "auto"
# セリフ・効果音の言語（auto は指示文の言語に合わせる）
PROMPT_EXPANDER_MANGA_TEXT_LANGUAGES: Final[tuple[str, ...]] = ("auto", "ja", "en")
DEFAULT_PROMPT_EXPANDER_MANGA_TEXT_LANGUAGE: Final[str] = "auto"
# 読み順（rtl = 日本式: 右上始まりで右→左・上→下、ltr = 西洋式）
PROMPT_EXPANDER_MANGA_READING_DIRECTIONS: Final[tuple[str, ...]] = ("rtl", "ltr")
DEFAULT_PROMPT_EXPANDER_MANGA_READING_DIRECTION: Final[str] = "rtl"


def max_character_prompts(image_model: str | None) -> int:
    """画像モデルに応じたキャラクタープロンプト上限を返す。"""
    return (
        MAX_CHARACTER_PROMPTS_V5
        if is_v5_image_model(image_model)
        else MAX_CHARACTER_PROMPTS_V45
    )


def max_character_prompts_map() -> dict[str, int]:
    """選択可能モデルごとの上限（設定応答に同梱する）。"""
    return {
        model: max_character_prompts(model)
        for model in PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS
    }


def is_prompt_expander_image_model(name: str | None) -> bool:
    return bool(name) and name in PROMPT_EXPANDER_IMAGE_MODEL_OPTIONS


def supports_manga_mode(image_model: str | None) -> bool:
    """漫画モード（コマ割り・文字描画）は V5 系モデルのみ対応。"""
    return is_v5_image_model(image_model)


def supports_precise_reference(image_model: str | None) -> bool:
    """精密参照（character reference）は V4.5 系モデルのみ対応（V5 は API 非対応）。"""
    return is_prompt_expander_image_model(image_model) and not is_v5_image_model(
        image_model
    )


def normalize_reference_type(value: object) -> str:
    """保存値・入力値を参照種別のいずれかに丸める（不正値は既定値）。"""
    return (
        value  # type: ignore[return-value]
        if isinstance(value, str) and value in PROMPT_EXPANDER_REFERENCE_TYPES
        else DEFAULT_PROMPT_EXPANDER_REFERENCE_TYPE
    )


def emphasize_tag(tag: str, level: int) -> str:
    """NovelAI の {} 記法でタグを強調する。level 0 は素通し。"""
    if level <= 0:
        return tag
    return "{" * level + tag + "}" * level


def normalize_transparent_emphasis(value: object) -> int:
    """保存値・入力値を強調レベルの範囲へ丸める（不正値は既定値）。"""
    if isinstance(value, bool) or not isinstance(value, int):
        return DEFAULT_PROMPT_EXPANDER_TRANSPARENT_EMPHASIS
    if value not in PROMPT_EXPANDER_TRANSPARENT_EMPHASIS_LEVELS:
        return DEFAULT_PROMPT_EXPANDER_TRANSPARENT_EMPHASIS
    return value


def transparent_background_tags(
    image_model: str | None, emphasis: int = 0
) -> tuple[str, ...]:
    """背景透過のために正プロンプト末尾へ足すタグ（モデル世代で分岐）。

    emphasis は V4.5 系の背景タグにだけ効く。V5 はネイティブ透過なので常に無強調。
    """
    if is_v5_image_model(image_model):
        return TRANSPARENT_BACKGROUND_TAGS_V5
    level = max(0, emphasis)
    if level == 0:
        return TRANSPARENT_BACKGROUND_TAGS_V45
    return tuple(
        emphasize_tag(tag, level)
        if tag in TRANSPARENT_BACKGROUND_EMPHASIZED_TAGS
        else tag
        for tag in TRANSPARENT_BACKGROUND_TAGS_V45
    )


def normalize_manga_panel_count(value: object) -> int:
    """保存値・入力値を 0（おまかせ）または範囲内の整数に丸める。"""
    if isinstance(value, bool) or not isinstance(value, int):
        return PROMPT_EXPANDER_MANGA_PANEL_COUNT_AUTO
    if value < PROMPT_EXPANDER_MANGA_PANEL_COUNT_MIN:
        return PROMPT_EXPANDER_MANGA_PANEL_COUNT_AUTO
    return min(value, PROMPT_EXPANDER_MANGA_PANEL_COUNT_MAX)
